libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c

```diff
@@ -3689,14 +3689,40 @@ CK_DECLARE_FUNCTION( CK_RV, C_Sign )( CK_SESSION_HANDLE hSession,
             {
                 if( pdTRUE == xSemaphoreTake( pxSessionObj->xSignMutex, portMAX_DELAY ) )
                 {
-                    lMbedTLSResult = mbedtls_pk_sign( &pxSessionObj->xSignKey,
-                                                      MBEDTLS_MD_NONE,
-                                                      pData,
-                                                      ulDataLen,
-                                                      pxSignatureBuffer,
-                                                      &xExpectedInputLength,
-                                                      mbedtls_ctr_drbg_random,
-                                                      &xP11Context.xMbedDrbgCtx );
+                    /* Per mbed TLS documentation, if using RSA, md_alg should
+                     * be MBEDTLS_MD_NONE. If ECDSA, md_alg should never be
+                     * MBEDTLS_MD_NONE. SHA-256 will be used for ECDSA for
+                     * consistency with the rest of the port.
+                     */
+                    if( pxSessionObj->xOperationSignMechanism == CKM_RSA_PKCS )
```

Comment:
The logic in lines 3698-3707 and 3709-3718 is identical except for the value of the second parameter.  Also, the code as written falls through at line 3721 rather than asserting as it probably should.  This would be much easier to read as:

```c
/*
* NOTE: assert() here rather than allowing a silent fall-through; this code
* cannot handle any other signing type.
*/
assert( pxSessionObj->xOperationSignMechanism == CKM_ECDSA ||
        pxSessionObj->xOperationSignMechanism == CKM_RSA_PKCS );

uint32_t ulMessageDigestType = 
    pxSessionObj->xOperationSignMechanism == CKM_ECDSA ?
        MBEDTLS_MD_SHA256 : MBEDTLS_MD_NONE;

lMbedTLSResult = mbedtls_pk_sign( &pxSessionObj->xSignKey,
        ulMessageDigestType,
        .
        .
        .
```

Reply:
Good callout, I condensed the code to more efficient form based on your feedback